File: backend/app/models/chapter.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel


class ChapterCreate(BaseModel):
    chapter_number: int
    title: Optional[str] = None
    content: str
    status: str = "draft"        # draft | scheduled | published
    publish_at: Optional[datetime] = None  # future date for VIP scheduling


class ChapterUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    status: Optional[str] = None
    publish_at: Optional[datetime] = None


class ChapterListItem(BaseModel):
    id: str
    novel_id: str
    chapter_number: int
    title: Optional[str] = None
    word_count: int
    status: str
    publish_at: Optional[datetime] = None
    published_at: Optional[datetime] = None
    views: int
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


class ChapterContent(ChapterListItem):
    """Full chapter with content, plus navigation."""
    content: str
    prev_chapter: Optional[int] = None   # chapter_number of previous
    next_chapter: Optional[int] = None   # chapter_number of next
    novel_title: Optional[str] = None


class ReadingProgress(BaseModel):
    user_id: str
    novel_id: str
    last_chapter_read: int
    chapters_read_list: list[int]
    updated_at: datetime
    model_config = {"from_attributes": True}

File: frontend/lib/types/chapter.ts
export interface ChapterListItem {
  id: string;
  novel_id: string;
  chapter_number: number;
  title: string | null;
  word_count: number;
  status: "draft" | "scheduled" | "published";
  publish_at: string | null;
  published_at: string | null;
  views: number;
  created_at: string;
  updated_at: string;
}

export interface ChapterContent extends ChapterListItem {
  content: string;
  prev_chapter: number | null;
  next_chapter: number | null;
  novel_title: string | null;
}

export interface ReadingProgress {
  user_id: string;
  novel_id: string;
  last_chapter_read: number;
  chapters_read_list: number[];
  updated_at: string;
}

export interface LibraryItem {
  novel_id: string;
  last_chapter_read: number;
  chapters_read_list: number[];
  updated_at: string;
  novel: {
    id: string;
    title: string;
    author: string;
    cover_url: string | null;
    status: string;
    total_chapters: number;
    updated_at: string;
  } | null;
}
